File: 0267-Palindrome Permutation II/0267-Palindrome Permutation II.cpp
class Solution {
public:
    vector<string> generatePalindromes(string s) {
        vector<int> table(128);
        for (char c : s) {
            ++table[c];
        }

        vector<string> result;
        int odd = 0;
        string mid;
        string seed;
        for (int i = 0; i < table.size(); ++i) {
            if (table[i] % 2) {
                ++odd;
                mid += i;
                if (odd > 1) {
                    return result;
                }
            }
            table[i] >>= 1;
            seed += string(table[i], i);
        }
        
        string path;
        vector<bool> used(seed.size());
        dfs(seed, path, mid, used, result);
        return result;
    }

private:
    void dfs(string& seed, string& path, string mid, vector<bool>& used, vector<string>& result) {
        if (path.size() == seed.size()) {
            string rev(path);
            reverse(rev.begin(), rev.end());
            result.push_back(path + mid + rev);
            return;
        }

        for (int i = 0; i < seed.size(); ++i) {
            if (used[i]) {
                continue;
            }
            if (i > 0 && seed[i] == seed[i - 1] && !used[i - 1]) {
                continue;
            }
            used[i] = true;
            path.push_back(seed[i]);
            dfs(seed, path, mid, used, result);
            path.pop_back();
            used[i] = false;
        }
    }
};

File: 0201-0300/0267-Palindrome Permutation II/0267-Palindrome Permutation II.py
import collections
class Solution:
    def generatePalindromes(self, s: str) -> List[str]:
        table = collections.Counter(s)
        odd = 0
        mid = ['']
        seed = ''
        for k, v in table.items():
            if v & 1:
                odd += 1                
                if odd > 1:
                    return ''
                mid = [k]
            seed += k * (v // 2)

        path = []
        result = []
        used = [False] * len(seed)
        def dfs():
            if len(path) == len(seed):
                result.append(''.join(path + mid + path[::-1]))
            for i in range(len(seed)):
                if used[i]:
                    continue
                if i > 0 and seed[i] == seed[i - 1] and not used[i - 1]:
                    continue
                used[i] = True
                path.append(seed[i])
                dfs()
                path.pop()
                used[i] = False
        
        dfs()
        return result
